### PR TITLE
Script for client-regeneration simulation

### PR DESCRIPTION
**Pull request by ANich**
Script version of the [client-regeneration notebook](https://colab.research.google.com/drive/1-VJr7jHMnjJ1UVQ2leXtBC6wayrCKMgh#scrollTo=yLSUPsa4OBC3&uniqifier=1) that we can run on Airflow

To run from client-regeneration dir:

```
docker build -t client-generation .
docker run client-regeneration python client_regeneration/main.py --seed 10
```

### CODE REVIEW

File: jobs/client-regeneration/client_regeneration/main.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+    -- )
+
+    -- remove the clients used as replacements from the churn pool (we only want them to serve as one client's replacement)
+    DELETE FROM `mozdata.analysis.regen_sim_churn_pool_{str(seed)}`
+    WHERE client_id IN (
+        SELECT replacement_id
+        FROM ( SELECT replacement_id
+          FROM `mozdata.analysis.regen_sim_client_replacements_{str(seed)}`
+          WHERE regen_date = "{str(date)}" )
+    );
```
Comment (ANich) [inline jobs/client-regeneration/client_regeneration/main.py:135]:
```suggestion
          WHERE regen_date = "{str(date)}" )
          AND replacement_id IS NOT NULL
    );
```

File: jobs/client-regeneration/client_regeneration/main.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+
+    -- find any cases where the regenerated IDs are also in the churn pool - replace them with their sampled replacement client.
+    UPDATE `mozdata.analysis.regen_sim_churn_pool_{str(seed)}` c
+    SET client_id = r.replacement_id
+    FROM
+      (SELECT client_id, replacement_id
+      FROM `mozdata.analysis.regen_sim_client_replacements_{str(seed)}`
+      WHERE regen_date = "{str(date)}" ) r
```
Comment (ANich) [inline jobs/client-regeneration/client_regeneration/main.py:143]:
```suggestion
      WHERE regen_date = "{str(date)}" 
      AND replacement_id IS NOT NULL ) r
```

File: jobs/client-regeneration/client_regeneration/main.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+    help="Date to start looking for replacements and writing history.",
+    default=DEFAULT_START_DATE,
+)
+@click.option(
+    "--end_date",
+    required=True,
+    type=click.DateTime(),
+    help="Date to stop looking for replacements and writing history.",
+    default=DEFAULT_END_DATE,
+)
```
Comment (ANich) [inline jobs/client-regeneration/client_regeneration/main.py:368]: TODO: Add lookback and column list as parameters?

File: .circleci/config.yml [hunk 4]
```diff
@@ -77,6 +77,22 @@ jobs:
...
+      - compare-branch:
+          pattern: ^jobs/client-regeneration/
+      - setup_remote_docker:
+          version: << pipeline.parameters.docker-version >>
+      - run:
+          name: Build Docker image
+          command: docker build -t app:build jobs/client-regeneration/
+      - run:
+          name: Test Code
+          command: docker run app:build pytest --flake8 --black
```
Comment (ANich) [inline .circleci/config.yml:94]: There aren't actually any tests we can probably remove this

File: jobs/client-regeneration/client_regeneration/main.py [hunk 5]
```diff
@@ -0,0 +1,383 @@
+import click
+
+from google.cloud import bigquery
+from datetime import datetime, timedelta
+
+DEFAULT_START_DATE = "2022-04-01"
+DEFAULT_END_DATE = "2022-06-01"
+
+COLUMN_LIST = [
```
Comment (irrationalagent) [inline jobs/client-regeneration/client_regeneration/main.py:9]: let's just use these columns for now. 

```suggestion
COLUMN_LIST = [
   "country",
    "device_model",
    "device_manufacturer",
]
```

File: jobs/client-regeneration/client_regeneration/main.py [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+    job.result()
+
+
+def init_regen_pool(client, seed, start_date):
+    # init a new version of the regen pool, write table to BQ and name it with the random seed that will be used in
+    # the sampling
+    # TODO: Can we get rid of this DECLARE?
+    q = f"""DECLARE start_date DATE DEFAULT DATE("{start_date}"); CREATE OR REPLACE TABLE 
+    mozdata.analysis.regen_sim_regen_pool_{seed} PARTITION BY regen_date AS SELECT * FROM 
+    mozdata.analysis.regen_sim_regen_pool_v1 WHERE regen_date >= start_date;"""
```
Comment (irrationalagent) [inline jobs/client-regeneration/client_regeneration/main.py:44]: i made a v2 version of the source table that goes back further, let's use that

```suggestion
    mozdata.analysis.regen_sim_regen_pool_v2 WHERE regen_date >= start_date;"""
```

File: jobs/client-regeneration/client_regeneration/main.py [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+
+    job = client.query(q)
+    job.result()
+
+
+def init_churn_pool(client, seed, start_date, lookback):
+    # init a new version of the churn pool, write table to BQ and name it with the random seed that will be used in
+    # the sampling
+    q = f"""CREATE OR REPLACE TABLE mozdata.analysis.regen_sim_churn_pool_{seed} PARTITION BY last_reported_date AS 
+    SELECT * FROM mozdata.analysis.regen_sim_churn_pool_v1 WHERE last_reported_date >= DATE_SUB(DATE("{start_date}"),
```
Comment (irrationalagent) [inline jobs/client-regeneration/client_regeneration/main.py:54]:
```suggestion
    SELECT * FROM mozdata.analysis.regen_sim_churn_pool_v2 WHERE last_reported_date >= DATE_SUB(DATE("{start_date}"), 
```

File: jobs/client-regeneration/client_regeneration/main.py [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+        AND c.submission_date >= DATE("{start_date}")
+    """
+
+    job = client.query(q)
+    job.result()
+
+    return table_name
+
+
+def write_usage_history(client, seed, start_date):
```
Comment (irrationalagent) [inline jobs/client-regeneration/client_regeneration/main.py:212]: this function will probably change

File: jobs/client-regeneration/client_regeneration/main.py [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+    create_replacements(
+        client,
+        seed=seed,
+        start_date=start_date,
+        end_date=end_date,
+        column_list=column_list,
+        lookback=lookback,
+    )
+    # next part isn't done yet  # TODO: This works now right?
+    write_attributed_clients_history(client, seed=seed, start_date=start_date)
```
Comment (irrationalagent) [inline jobs/client-regeneration/client_regeneration/main.py:349]: we can comment out these two lines

File: jobs/client-regeneration/client_regeneration/main.py [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+
+COLUMN_LIST = [
+    "adjust_campaign",
+    "adjust_ad_group",
+    "adjust_creative",
+    "adjust_network",
+    "country",
+    "device_model",
+    "device_manufacturer",
+]
```
Comment (irrationalagent) [inline jobs/client-regeneration/client_regeneration/main.py:17]: let's just use these columns for now

```suggestion
COLUMN_LIST = [
    "adjust_campaign",
    "adjust_ad_group",
    "adjust_creative",
    "adjust_network",
    "country",
    "device_model",
    "device_manufacturer",
]
```

File: jobs/client-regeneration/client_regeneration/main.py [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+        client,
+        seed=seed,
+        start_date=start_date,
+        end_date=end_date,
+        column_list=column_list,
+        lookback=lookback,
+    )
+    # next part isn't done yet  # TODO: This works now right?
+    write_attributed_clients_history(client, seed=seed, start_date=start_date)
+    write_usage_history(client, seed=seed, start_date=start_date)
```
Comment (irrationalagent) [inline jobs/client-regeneration/client_regeneration/main.py:350]: these are likely to change, we can comment them for now.

```suggestion
    # write_attributed_clients_history(client, seed=seed, start_date=start_date)
    # write_usage_history(client, seed=seed, start_date=start_date)
```

**Comment by irrationalagent** [comment]
let's avoid using seed `123` in testing this because that will overwrite some test data that im analyzing.

**Review by irrationalagent — APPROVED** [review]
the main.py looks good to me, not familiar with docker enough to review the boilerplate

**Pull request merged**